### PR TITLE
Update React 16->17, TypeScript, and some other node deps

### PR DESCRIPTION
**Pull request by juliusv**
This updates React, TypeScript, and some other node packages (but not
everything).

A couple of notes:

- `enzyme-adapter-react-16` does not have a React 17 equivalent yet, so I
  switched to the fork `@wojtekmaj/enzyme-adapter-react-17`
- A bunch of tests are still failing because I think in the enzyme testing
  environment, a browser API (`ResizeObserver`) is missing, and maybe for other
  reasons. This needs to be explored + fixed.
- The TypeScript update introduced more stringent rules, which required fixing
  up a bunch of pieces of code a bit.
- The `use-media` package doesn't work with React 17 yet, so I just built our
  own minimal `useMedia` hook instead (just a couple of lines).
- I commented out part of the code in `withStartingIndicator.tsx` because it
  fails the now-stricter lint checks. It needs to be fixed (and not commented
  out).

Signed-off-by: Julius Volz <julius.volz@gmail.com>

### CODE REVIEW

**Comment by juliusv** [comment]
@Nexucis This is how far I got with the React + TypeScript + etc. updates before vacation tomorrow. The notes in the PR body describe what I did so far and what is still missing (and probably some other stuff can be cleaned up). In case you would like to continue working on it (so you are not blocked for the other things waiting on this), feel free to push further commits onto my branch (but not squash mine :)), otherwise I will pick it up again later.

**Comment by Nexucis** [comment]
thanks @juliusv for your work!! I will try to continue to work on it :)
Let's see how far I can go !

**Comment by Nexucis** [comment]
> I commented out part of the code in withStartingIndicator.tsx because it
fails the now-stricter lint checks. It needs to be fixed (and not commented
out).

This is fixed

> A bunch of tests are still failing because I think in the enzyme testing
environment, a browser API (ResizeObserver) is missing, and maybe for other
reasons. This needs to be explored + fixed.

Partially fixed. I have others errors now related to the canvas rendering and invalid dimensions for a plot.
* Link for the fix related to `ResizeObseerver` : https://github.com/maslianok/react-resize-detector/issues/145
* Link for fix related to the canvas issue : https://stackoverflow.com/questions/33269093/how-to-add-canvas-support-to-my-tests-in-jest

**Comment by Nexucis** [comment]
mmm I'm going to split PR in two parts.

This one will remain for upgrading react17. Will create another one where we "just" upgrade `react-script`. Like that we should be able split the work and be sure where the errors are coming from.

I'm a bit lost here on where the errors are coming from and it's hard to solve them.

**Comment by Nexucis** [comment]
I created the PR https://github.com/prometheus/prometheus/pull/9296 that upgrades the `react-script`.
After the upgrade some tests failed and I fixed them.

I will do another PR after the one above is merged to upgrade typescript.
Once it is done, then back to this one to finally taking care only on upgrading React to the last version

**Comment by Nexucis** [comment]
oh yeah the diff is so much better now.
Now we can see the changes from upgrading react only and not because of the upgrade react-script and eslint

**Comment by roidelapluie** [comment]
`TypeError: window.matchMedia is not a function`

**Comment by Nexucis** [comment]
@roidelapluie yeah this one will take more time and definitely shouldn't be considered for the release tomorrow

**Comment by Nexucis** [comment]
```
FAIL  src/App.test.tsx
  ● Test suite failed to run

    TypeError: window.matchMedia is not a function
```

this one is quite easy at it is required to take more code from the lib `useMedia` that contains the appropriate mock to be used during the tests.

```
FAIL  src/pages/graph/Graph.test.tsx
  ● Console

    console.error
      Error: Uncaught [Error: Invalid dimensions for plot, width = 0, height = 0]
```

This one is the hardest. I had to include a mock of the canvas by installing the dependency `jest-mock-canvas`. (no idea why it wasn't required before).  And now we have this issue and no idea how to solve it

I'm wondering if it's because when fixing the `ResizeObserver`, I didn't use this part of the code proposed in the solution shared 

```typescript
wrapper = mount(
      <MyComponent />
    );
```

I didn't get what I am supposed to do with that. And also `wrapper` is not declared here, so I guess it should be declared somewhere else but where ? And how ? 

```
● Graph › plotSetAndDraw › should call spyPlotSetAndDraw on legend hover

    expect(jest.fn()).toHaveBeenCalledTimes(expected)
```
Finally this one was fixed in #9296 using the parameter of the `react-script` `--resetMocks=false`. But now it's failing again. So weird ....

**Comment by Nexucis** [comment]
> I'm wondering if it's because when fixing the `ResizeObserver`, I didn't use this part of the code proposed in the solution shared 

yeah ok nothing related. The piece of the code I didn't use, doesn't bring any additional things to fix the issue.

**Comment by Nexucis** [comment]
@juliusv it's ready for a review :). I can't assign it to you since you create the PR :joy:

File: web/ui/react-app/src/pages/graph/Graph.test.tsx [hunk 1]
```diff
@@ -9,6 +9,19 @@ describe('Graph', () => {
   beforeAll(() => {
     jest.spyOn(window, 'requestAnimationFrame').mockImplementation((cb: any) => cb());
   });
+
+  // fix coming from https://github.com/maslianok/react-resize-detector#testing-with-enzyme-and-jest
```
Comment (roidelapluie) [inline web/ui/react-app/src/pages/graph/Graph.test.tsx:13]:
```suggestion
  // Source: https://github.com/maslianok/react-resize-detector#testing-with-enzyme-and-jest
```

File: web/ui/react-app/src/pages/graph/Graph.test.tsx [hunk 2]
```diff
@@ -9,6 +9,19 @@ describe('Graph', () => {
...
+    window.ResizeObserver = jest.fn().mockImplementation(() => ({
+      observe: jest.fn(),
+      unobserve: jest.fn(),
+      disconnect: jest.fn(),
+    }));
+  });
+
+  afterEach(() => {
+    window.ResizeObserver = ResizeObserver;
+  });
```
Comment (roidelapluie) [inline web/ui/react-app/src/pages/graph/Graph.test.tsx:24]: don't we need an extra blank line?

Comment (Nexucis) [reply web/ui/react-app/src/pages/graph/Graph.test.tsx:24]: mmm I think if it was the case, then the linter would yell

Comment (juliusv) [reply web/ui/react-app/src/pages/graph/Graph.test.tsx:24]: This is just a place where a human would expect an empty line (not relevant to the linter though), so let's put one there.

**Comment by juliusv** [comment]
:+1: Awesome, looks good to merge after that final nit comment.

File: web/ui/react-app/src/pages/graph/Graph.test.tsx (hunk 2 above)
Comment (Nexucis) [reply web/ui/react-app/src/pages/graph/Graph.test.tsx:24]: ah yeah ok, I misunderstood it was for code sanity. Yeah make sense.

**Comment by Nexucis** [comment]
it's done @juliusv

**Comment by juliusv** [comment]
:+1: Thanks, will merge after CI goes green :)

**Pull request merged**